x/ecocredit/server/core/msg_server.go

```diff
@@ -0,0 +1,1200 @@
+	"github.com/regen-network/regen-ledger/x/ecocredit/v1beta1"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// TODO: Revisit this once we have proper gas fee framework.
+// Tracking issues https://github.com/cosmos/cosmos-sdk/issues/9054, https://github.com/cosmos/cosmos-sdk/discussions/9072
+const gasCostPerIteration = uint64(10)
+
+// CreateClass creates a new class of ecocredit
+//
+// The admin is charged a fee for creating the class. This is controlled by
+// the global parameter CreditClassFee, which can be updated through the
+// governance process.
+func (s serverImpl) CreateClass(ctx context.Context, req *v1beta1.MsgCreateClass) (*v1beta1.MsgCreateClassResponse, error) {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	// Charge the admin a fee to create the credit class
+	adminAddress, err := sdk.AccAddressFromBech32(req.Admin)
+	if err != nil {
+		return nil, err
+	}
+
+	var params ecocredit.Params
+	s.paramSpace.GetParamSet(sdkCtx, &params)
+	if params.AllowlistEnabled && !s.isCreatorAllowListed(sdkCtx, params.AllowedClassCreators, adminAddress) {
+		return nil, sdkerrors.ErrUnauthorized.Wrapf("%s is not allowed to create credit classes", adminAddress.String())
+	}
+
+	err = s.chargeCreditClassFee(sdkCtx, adminAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	creditType, err := s.getCreditType(sdkCtx, req.CreditTypeName)
+	if err != nil {
+		return nil, err
+	}
+
+	classSeq, err := s.getClassSequenceNo(ctx, req.CreditTypeName)
+	if err != nil {
+		return nil, fmt.Errorf("error getting class sequence")
+	}
+	classID := ecocredit.FormatClassID(creditType.Abbreviation, classSeq)
+
+	rowID, err := s.stateStore.ClassInfoStore().InsertReturningID(ctx, &ecocreditv1beta1.ClassInfo{
+		Name:       classID,
+		Admin:      adminAddress,
+		Metadata:   req.Metadata,
+		CreditType: req.CreditTypeName,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, issuer := range req.Issuers {
+		issuer, _ := sdk.AccAddressFromBech32(issuer)
+		if err = s.stateStore.ClassIssuerStore().Insert(ctx, &ecocreditv1beta1.ClassIssuer{
+			ClassId: classID,
+			Issuer:  issuer,
+		}); err != nil {
+			return nil, err
+		}
+	}
+
+	err = sdkCtx.EventManager().EmitTypedEvent(&v1beta1.EventCreateClass{
+		RowId:   rowID,
+		ClassId: classID,
+		Admin:   req.Admin,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &v1beta1.MsgCreateClassResponse{RowId: rowID, ClassId: classID}, nil
+}
+
+// CreateProject creates a new project.
+func (s serverImpl) CreateProject(ctx context.Context, req *v1beta1.MsgCreateProject) (*v1beta1.MsgCreateProjectResponse, error) {
+	sdkCtx := types.UnwrapSDKContext(ctx)
+	classID := req.ClassId
+	classInfo, err := s.stateStore.ClassInfoStore().GetByName(ctx, classID)
+	if err != nil {
+		return nil, err
+	}
+
+	err = s.assertClassIssuer(ctx, classID, req.Issuer)
+	if err != nil {
+		return nil, err
+	}
+
+	projectID := req.ProjectId
+	if projectID == "" {
+		found := false
+		for !found {
```

Comment:
why do we need a loop here? something seems not quite right here

Reply:
it was actually your suggestion in the original projects PR 😄 - https://github.com/regen-network/regen-ledger/pull/647#discussion_r766151278

Reply:
we need to loop to avoid a clash but the way this loop is structured is just confusing

Reply:
do you have any suggestions to make it clearer? i had thought it was pretty straightforward but open to suggestions